edb/edgeql/compiler/func.py

```diff
@@ -610,13 +621,40 @@ def compile_call_args(
                 f'${aname} of function {funcname}',
                 context=arg.context)
 
+        arg_ctxs[arg_ir] = arg_ctx
         kwargs[aname] = (arg_type, arg_ir)
 
-    return args, kwargs
+    return args, kwargs, arg_ctxs
+
+
+def process_path_log(arg_ctx: Optional[context.ContextLevel],
+                     arg_scope: Optional[irast.ScopeTreeNode]) -> None:
+    if arg_ctx and arg_ctx.path_log is not None:
+        # Since we don't know whether arguments are OPTIONAL or SET OF
+        # until after doing polymorphic matching, paths in an OPTIONAL
+        # or SET OF argument could get factored into an existing
+        # optional node, destroying its optionality.
+        #
+        # To fix this, we track all of the paths attached to the tree
+        # while compiling an argument and find and adjust the
+        # optionality of any factored out nodes after the fact.
+        for path_id in arg_ctx.path_log:
+            for prefix in path_id.iter_prefixes(include_ptr=True):
+                assert arg_scope is not None
+                # If the node is still here, nothing to do
+                desc = arg_scope.find_descendant(prefix)
+                if desc:
+                    continue
+                visible = arg_scope.find_visible(prefix)
+                if visible:
+                    if visible.optional_count:
+                        visible.optional_count -= 1
```

Comment:
Is it possible for `visible.optional_count` to be > 1 here?

Reply:
Yeah. If we have something like `SELECT (Foo.num ?= 0, Foo.num ?= Foo.num)`, the optional_count will go up to 2 and then come back down to 0. If we have ``SELECT (Foo.num ?= 0, Foo.num = Foo.num)`, it will go up to 2 and stay there.